[SR-5624] Fix to bring "dependencies" key to the last order of output json

hotchemi opened the pull request:
## Issue

- https://bugs.swift.org/browse/SR-5624

## Overview

This PR includes the following attempt.

- Fix to bring "dependencies" key to the last order of output json.
  - Fix to inject a function which determines order of contents keys of JSON when writing the contents to `OutputByteStream`.
  - Define a specific function to alter the json order in `JsonDumper`.
  - Add unit test case for the change.

xwu commented on Sources/Basic/JSON.swift:
```diff
@@ -92,18 +92,18 @@ extension JSON: Equatable {
 extension JSON {
     /// Encode a JSON item into a string of bytes.
-    public func toBytes(prettyPrint: Bool = false) -> ByteString {
+    public func toBytes(prettyPrint: Bool = false, contentsKeyOrder: (String, String) -> Bool = {a, b in a < b}) -> ByteString {
```
The closure can be written as `{ $0 < $1 }`. Following Swift conventions, the argument label should be named something like `sortingKeysBy` and the internal parameter should be`areInIncreasingOrder`.

xwu commented on Sources/Commands/show-dependencies.swift:
```diff
@@ -107,7 +107,13 @@ private final class JSONDumper: DependenciesDumper {
+
+    /// To bring `dependencies` key to the last order.
+    /// See: https://bugs.swift.org/browse/SR-5624
+    func contentsKeyOrder(a: String, b: String) -> Bool {
```
There is no need for this to be a method; it can be a closure above.

hotchemi replied:
I deliberately define as a method to improve a searchability.

xwu commented on Sources/Commands/show-dependencies.swift:
```diff
@@ -107,7 +107,13 @@ private final class JSONDumper: DependenciesDumper {
+    /// To bring `dependencies` key to the last order.
+    /// See: https://bugs.swift.org/browse/SR-5624
+    func contentsKeyOrder(a: String, b: String) -> Bool {
+        return b == "dependencies" ? a > b : a < b
```
This appears to be incorrect as a sort predicate and only happens to work because of the particular keys hardcoded above. Consider what would happen if you had keys named "data", "dependencies", and "name".

A sort predicate that unconditionally moves `dependencies` last would be `{ $0 != "dependencies" || $1 == "dependencies" || $0 < $1 }`.

xwu commented on Tests/BasicTests/JSONTests.swift:
```diff
@@ -26,6 +26,21 @@ class JSONTests: XCTestCase {
+    func testEncodingWithContentsKeyOrder() {
+        // Test encoding with customized contents key order.
+        func encode(_ item: JSON) -> String {
+            /// See: https://bugs.swift.org/browse/SR-5624
```
Nit: this should be a regular `//` comment, not a doc comment.

xwu commented on Sources/Basic/JSON.swift:
```diff
@@ -142,9 +142,8 @@ extension JSON: ByteStreamable {
             }
             stream <<< (shouldIndent ? "\n" : "") <<< indentStreamable() <<< "]"
         case .dictionary(let contents):
-            // We always output in a deterministic order.
```
Why delete this comment?

hotchemi replied:
Because now it's not always deterministic.

xwu replied:
In what way is it non-deterministic?

xwu commented on Sources/Basic/JSON.swift:
```diff
@@ -92,18 +92,18 @@ extension JSON: Equatable {
 extension JSON {
     /// Encode a JSON item into a string of bytes.
-    public func toBytes(prettyPrint: Bool = false) -> ByteString {
+    public func toBytes(prettyPrint: Bool = false, sortingKeysBy: (String, String) -> Bool = { $0 < $1 }) -> ByteString {
```
Nit: for consistency, could you use the internal parameter name here too (as in, `sortingKeysBy areInIncreasingOrder: (String, String) -> Bool`)?

xwu commented on Sources/Basic/JSON.swift:
```diff
@@ -92,18 +92,18 @@ extension JSON: Equatable {
     /// Encode a JSON item into a JSON string
-    public func toString(prettyPrint: Bool = false) -> String {
-        guard let contents = self.toBytes(prettyPrint: prettyPrint).asString else {
+    public func toString(prettyPrint: Bool = false, sortingKeysBy: (String, String) -> Bool = { $0 < $1 }) -> String {
```
Nit: also here: `sortingKeysBy areInIncreasingOrder: (String, String) -> Bool`.

xwu commented on Sources/Commands/show-dependencies.swift:
```diff
@@ -107,7 +107,13 @@ private final class JSONDumper: DependenciesDumper {
+
+    /// To bring `dependencies` key to the last order.
+    /// See: https://bugs.swift.org/browse/SR-5624
+    func sortingKeysBy(a: String, b: String) -> Bool {
```
Again, this should ideally not be a member function; nothing about it has anything to do with a particular instance of `JSONDumper`.

The way I see it, it can be moved above into `dump(dependenciesOf:)`, either as a closure or as its own function (like `convert(_:)` above). If you're going to choose to have it as its own function, please use a more descriptive name than just `sortingKeysBy` (something like `areInIncreasingOrderExceptDependenciesLast`--except maybe better; the name should generally begin with `is` or `are` because it returns a `Bool`).

But in either case, it should not be a member function of `JSONDumper`.

xwu commented on Tests/BasicTests/JSONTests.swift:
```diff
@@ -26,6 +26,21 @@ class JSONTests: XCTestCase {
     }
+
+    func testEncodingWithContentsKeyOrder() {
+        // Test encoding with customized contents key order.
```
Can you write a test that uses something other than the specific keys above? For instance, something that shows that the predicate I suggested is correct but wouldn't work with your original predicate?

xwu commented on Sources/Commands/show-dependencies.swift:
```diff
@@ -107,7 +107,13 @@ private final class JSONDumper: DependenciesDumper {
         }
 
-        print(convert(rootpkg).toString(prettyPrint: true))
+        print(convert(rootpkg).toString(prettyPrint: true, sortingKeysBy: sortingKeysBy(a:b:)))
```
Nit, for clarity and ease of reading, you can omit the `(a:b:)`, I believe.

hotchemi commented:
@xwu thx for dedicated review but I'd like to wait core maintainer's review because I wonder I've been going to the wrong direction against core developer's thought.

hotchemi commented:
@aciidb0mb3r sorry for bother you, could you review just in case? I'd like to get your thought before addressing review points by @xwu . Thank you! 🙇

aciidb0mb3r commented:
@hotchemi I am out on vacation, I'll review this ASAP when I return tomorrow!

aciidb0mb3r commented:
@hotchemi Thank you for working on this. I think it will be better if we introduced a sorted dictionary structure in the JSON enum, so clients can properly control the order. What do you think?

hotchemi commented:
@aciidb0mb3r thx for your quick review 🙇
Actually I'm fine with it and it'd be much more flexible way to inject an order compare algorithm!
Before go on it, could I get more concrete example to let us being on the same page?

I still don't imagine entire picture in my mind, but I found `SortedDictionary` in [C#](https://msdn.microsoft.com/en-us/library/f7fta44c(v=vs.110).aspx) doc.

So in short:
- `SortedDictionary` extends `Dictionary` in standard lib
- Usr can inject kind of [Comparer](https://msdn.microsoft.com/en-us/library/8ehhxeaf(v=vs.110).aspx)(or closure) on initializer
- Use it in Json dictionary

Am I correct?

aciidb0mb3r commented:
Ah, I actually meant `OrderedDictionary` and not `SortedDictionary` (sorry for the confusion).
Basically, I think the JSON output should be same as the one set during toJSON method. So if you we have a custom struct Foo, we should be able to do something like this:

```swift
struct Foo: JSONSerializable {
    func toJSON() -> JSON {
         return .orderedDictionary([
            "key2": "value2",
            "key1": "value1",
        ])
    }
}
```

It looks like Swift already has a type called `DictionaryLiteral` for creating ordered dictionary but it does not support fast lookups. https://developer.apple.com/documentation/swift/dictionaryliteral
I don't think we need fast lookups for our use case, so we might be able to directly use `DictionaryLiteral` for now. To do so, just add another case in JSON enum:

```swift
enum JSON {
    ...
    case dictionaryLiteral(DictionaryLiteral)
}
```

hotchemi commented:
Oh I missed `DictionaryLiteral` actually...then the change would be so much simple!
Let me close the PR and create another one 🙇
